File: 剑指 Offer 32 - I. 从上到下打印二叉树.py
'''
Description: 
Version: 1.0
Autor: Vicro
Date: 2021-01-05 14:01:16
LastEditTime: 2021-01-05 14:04:31
FilePath: \Leetcode\剑指 Offer 32 - I. 从上到下打印二叉树.py
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import collections


"""
RESULT: Accept
TIME:     40ms    BEAT: 69.78%    O(n) = 
MEMORY: 14.9MB    BEAT:  7.85%    O(n) = 
Description: BFS
"""

class Solution:
    def levelOrder(self, root: TreeNode) -> List[int]:
        if not root: return []
        res, queue = [], collections.deque()
        queue.append(root)

        while queue:
            node = queue.popleft()
            res.append(node.val)
            if node.left: queue.append(node.left)
            if node.right: queue.append(node.right)
        
        return res

File: 剑指offer/剑指 Offer 32 - I. 从上到下打印二叉树.cpp
/*
 * @Description: 
 * @Version: 1.0
 * @Autor: 冰凝水
 * @Date: 2021-01-05 13:30:01
 * @LastEditTime: 2021-01-05 14:03:43
 * @FilePath: \Leetcode\剑指 Offer 32 - I. 从上到下打印二叉树.cpp
 */

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

#include <vector>
#include <iostream>
#include <queue>
using namespace std;


/*
RESULT: Accept
TIME:      8ms    BEAT: 42.42%    O(n) = n
MEMORY: 12.4MB    BEAT: 21.83%    O(n) = n
USED TIME: 04:50
Description: BFS
*/

class Solution {
public:
    vector<int> levelOrder(TreeNode* root) {
        vector<int> res;
        if (root == nullptr) return {}; 
        
        queue<TreeNode*> Bfs;
        Bfs.push(root);
        while (!Bfs.empty())
        {
            TreeNode* temp = Bfs.front();
            res.push_back(temp->val);
            Bfs.pop();
            if (temp->left) Bfs.push(temp->left);
            if (temp->right) Bfs.push(temp->right);
        }
        return res;
    }
};